File: app/src/main/java/app/kamy/saatApp/domain/tools/ZakatCalculator.kt
package app.kamy.saatApp.domain.tools

import kotlin.math.floor
import kotlin.math.min

sealed interface ZakatCalculationResult

data class ZakatMaalCalculationResult(
    val zakatableWealth: Double,
    val nisabGoldGrams: Double,
    val nisabSilverGrams: Double,
    val nisabGoldValue: Double,
    val nisabSilverValue: Double,
    val effectiveNisab: Double,
    val zakatDue: Double,
    val meetsNisab: Boolean,
    val usedSilverNisab: Boolean
) : ZakatCalculationResult

data class ZakatFitrahCalculationResult(
    val familyMembers: Int,
    val stapleWeightPerPersonKg: Double,
    val staplePricePerKg: Double,
    val totalStapleKilograms: Double,
    val zakatDue: Double
) : ZakatCalculationResult

data class GoldPriceQuote(
    val goldPerGramIdr: Double,
    val silverPerGramIdr: Double,
    val sourceLabel: String,
    val fetchedAtMillis: Long
)

enum class ZakatType {
    MAAL,
    FITRAH
}

object ZakatCalculator {
    const val NISAB_GOLD_GRAMS = 85.0
    const val NISAB_SILVER_GRAMS = 595.0
    const val ZAKAT_RATE = 0.025
    const val FITRAH_WEIGHT_PER_PERSON_KG = 2.5
    private const val TROY_OZ_GRAMS = 31.1034768

    fun goldUsdPerGram(usdPerTroyOz: Double): Double = usdPerTroyOz / TROY_OZ_GRAMS

    fun calculate(
        cash: Double,
        goldGrams: Double,
        silverGrams: Double,
        investments: Double,
        debts: Double,
        goldPricePerGram: Double,
        silverPricePerGram: Double
    ): ZakatMaalCalculationResult {
        val goldValue = goldGrams * goldPricePerGram
        val silverValue = silverGrams * silverPricePerGram
        val net = (cash + goldValue + silverValue + investments - debts).coerceAtLeast(0.0)
        val nisabGoldValue = NISAB_GOLD_GRAMS * goldPricePerGram
        val nisabSilverValue = NISAB_SILVER_GRAMS * silverPricePerGram
        val effectiveNisab = min(nisabGoldValue, nisabSilverValue)
        val usedSilver = nisabSilverValue < nisabGoldValue
        val meetsNisab = net >= effectiveNisab
        val zakatDue = if (meetsNisab) floor(net * ZAKAT_RATE * 100) / 100.0 else 0.0
        return ZakatMaalCalculationResult(
            zakatableWealth = net,
            nisabGoldGrams = NISAB_GOLD_GRAMS,
            nisabSilverGrams = NISAB_SILVER_GRAMS,
            nisabGoldValue = nisabGoldValue,
            nisabSilverValue = nisabSilverValue,
            effectiveNisab = effectiveNisab,
            zakatDue = zakatDue,
            meetsNisab = meetsNisab,
            usedSilverNisab = usedSilver
        )
    }

    fun calculateFitrah(
        familyMembers: Int,
        staplePricePerKg: Double,
        stapleWeightPerPersonKg: Double = FITRAH_WEIGHT_PER_PERSON_KG
    ): ZakatFitrahCalculationResult {
        val totalWeight = familyMembers * stapleWeightPerPersonKg
        val zakatDue = floor(totalWeight * staplePricePerKg * 100) / 100.0
        return ZakatFitrahCalculationResult(
            familyMembers = familyMembers,
            stapleWeightPerPersonKg = stapleWeightPerPersonKg,
            staplePricePerKg = staplePricePerKg,
            totalStapleKilograms = totalWeight,
            zakatDue = zakatDue
        )
    }

    fun silverPriceFromGold(goldPricePerGram: Double): Double = goldPricePerGram / 80.0
}

File: iosApp/Saat/Domain/Models/ZakatCalculator.swift
//
//  ZakatCalculator.swift
//  Saat
//
//  Created by Elmee on 25/06/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

enum ZakatType: String, CaseIterable, Codable, Sendable {
    case maal = "MAAL"
    case fitrah = "FITRAH"
}

struct ZakatMaalCalculationResult: Codable, Sendable {
    let zakatableWealth: Double
    let nisabGoldGrams: Double
    let nisabSilverGrams: Double
    let nisabGoldValue: Double
    let nisabSilverValue: Double
    let effectiveNisab: Double
    let zakatDue: Double
    let meetsNisab: Bool
    let usedSilverNisab: Bool
}

struct ZakatFitrahCalculationResult: Codable, Sendable {
    let familyMembers: Int
    let stapleWeightPerPersonKg: Double
    let staplePricePerKg: Double
    let totalStapleKilograms: Double
    let zakatDue: Double
}

struct GoldPriceQuote: Codable, Sendable {
    let goldPerGramIdr: Double
    let silverPerGramIdr: Double
    let sourceLabel: String
    let fetchedAtMillis: Int64
}

enum ZakatCountry: String, CaseIterable, Codable, Sendable, Identifiable {
    case indonesia = "ID"
    case malaysia = "MY"
    case singapore = "SG"
    case brunei = "BN"
    
    var id: String { rawValue }
    
    var emoji: String {
        switch self {
        case .indonesia: return "🇮🇩"
        case .malaysia: return "🇲🇾"
        case .singapore: return "🇸🇬"
        case .brunei: return "🇧🇳"
        }
    }
    
    var labelKey: String {
        switch self {
        case .indonesia: return "zakat_country_indonesia"
        case .malaysia: return "zakat_country_malaysia"
        case .singapore: return "zakat_country_singapore"
        case .brunei: return "zakat_country_brunei"
        }
    }
}

struct ZakatCalculator {
    static let NISAB_GOLD_GRAMS = 85.0
    static let NISAB_SILVER_GRAMS = 595.0
    static let ZAKAT_RATE = 0.025
    static let FITRAH_WEIGHT_PER_PERSON_KG = 2.5
    private static let TROY_OZ_GRAMS = 31.1034768

    static func goldUsdPerGram(usdPerTroyOz: Double) -> Double {
        return usdPerTroyOz / TROY_OZ_GRAMS
    }

    static func calculate(
        cash: Double,
        goldGrams: Double,
        silverGrams: Double,
        investments: Double,
        debts: Double,
        goldPricePerGram: Double,
        silverPricePerGram: Double
    ) -> ZakatMaalCalculationResult {
        let goldValue = goldGrams * goldPricePerGram
        let silverValue = silverGrams * silverPricePerGram
        let net = max(0.0, cash + goldValue + silverValue + investments - debts)
        let nisabGoldVal = NISAB_GOLD_GRAMS * goldPricePerGram
        let nisabSilverVal = NISAB_SILVER_GRAMS * silverPricePerGram
        let effectiveNisab = min(nisabGoldVal, nisabSilverVal)
        let usedSilver = nisabSilverVal < nisabGoldVal
        let meetsNisab = net >= effectiveNisab
        let zakatDue = meetsNisab ? floor(net * ZAKAT_RATE * 100.0) / 100.0 : 0.0
        
        return ZakatMaalCalculationResult(
            zakatableWealth: net,
            nisabGoldGrams: NISAB_GOLD_GRAMS,
            nisabSilverGrams: NISAB_SILVER_GRAMS,
            nisabGoldValue: nisabGoldVal,
            nisabSilverValue: nisabSilverVal,
            effectiveNisab: effectiveNisab,
            zakatDue: zakatDue,
            meetsNisab: meetsNisab,
            usedSilverNisab: usedSilver
        )
    }

    static func calculateFitrah(
        familyMembers: Int,
        staplePricePerKg: Double,
        stapleWeightPerPersonKg: Double = FITRAH_WEIGHT_PER_PERSON_KG
    ) -> ZakatFitrahCalculationResult {
        let totalWeight = Double(familyMembers) * stapleWeightPerPersonKg
        let zakatDue = floor(totalWeight * staplePricePerKg * 100.0) / 100.0
        
        return ZakatFitrahCalculationResult(
            familyMembers: familyMembers,
            stapleWeightPerPersonKg: stapleWeightPerPersonKg,
            staplePricePerKg: staplePricePerKg,
            totalStapleKilograms: totalWeight,
            zakatDue: zakatDue
        )
    }

    static func silverPriceFromGold(goldPricePerGram: Double) -> Double {
        return goldPricePerGram / 80.0
    }
}
